jbmc/src/java_bytecode/convert_java_nondet.cpp

```diff
@@ -22,90 +22,141 @@ Author: Reuben Thomas, reuben.thomas@diffblue.com
 
 #include "java_object_factory.h" // gen_nondet_init
 
+/// Returns true if `expr` is a nondet pointer that isn't a function pointer or
+/// a void* pointer as these can be meaningfully non-det initalized. 
+static bool is_nondet_pointer(exprt expr)
+{
+  // If the expression type doesn't have a subtype then I guess it's primitive
+  // and we do not want to convert it. If the type is a ptr-to-void or a
+  // function pointer then we also do not want to convert it.
+  const typet &type = expr.type();
+  const bool non_void_non_function_pointer = type.id() == ID_pointer &&
+                                             type.subtype().id() != ID_empty &&
+                                             type.subtype().id() != ID_code;
+  return can_cast_expr<side_effect_expr_nondett>(expr) &&
```

Comment:
Unless you take the advice above then reversing the order of these would be more efficient as `&&` short-circuits.